### DISCUSSION TITLE
Cannot set a wire:model value to null using CustomEvent

### DISCUSSION THREAD

**Discussion by adriaanzon** [livewire/livewire#4617] [7. Bugs (archived)]
I am using Alpine to $dispatch an input event to set a wire:model to null. Since Livewire version 2.8.0, this isn't working anymore. This has to do with https://github.com/livewire/livewire/pull/4247, where the value supplied with the CustomEvent gets overwritten if it is "falsy". It gets overwritten with the value of the `value` attribute of the event's target. I've also explained it here: https://github.com/livewire/livewire/pull/4460#issuecomment-1019483088.

I have illustrated the problem in this video:

https://user-images.githubusercontent.com/4326420/152659164-78f15310-af3b-4a01-8525-9a95058d2264.mov

In the first example, I used button elements. When clicking the reset button, [e.target.value](https://github.com/livewire/livewire/pull/4247/files#diff-292ab00c07366b14c1c74551be0095ad4b3352f6ade951b1dbbc0621ccccd3a1R88) is equal to an empty string, because button elements can have a `value` attribute, so the DOM returns an empty string when you try accessing the button's value. 

In the second example, I used span elements instead of buttons. Here, it seems like nothing is happening when clicking the reset button. In reality, the `null` value from the input event is being overwritten with `undefined`, because span elements normally don't have a `value` attribute, so the DOM returns `undefined` when you try accessing the span's value. Because the value is undefined, it won't be present on the syncInput request, which will be ignored [here](https://github.com/livewire/livewire/pull/4460/files#diff-443b4dbefdc801e006721d5649042189c77fbcebc0c374f64ea1f58c10a5e9e3R18). That's why it seems like it isn't responding when clicking on the reset button.

I have uploaded the example to [this repository](https://github.com/adriaanzon/livewire-dispatch-bug-demo), where you can view the [component class](https://github.com/adriaanzon/livewire-dispatch-bug-demo/blob/main/app/Http/Livewire/LivewireDispatchBugDemo.php) and the [component template](https://github.com/adriaanzon/livewire-dispatch-bug-demo/blob/main/resources/views/livewire/livewire-dispatch-bug-demo.blade.php).

**Comment by jschwendener** [comment]
We're experiencing the same issue with custom Blade components which use `$dispatch` to update a `wire:model` bound attribute.

In our case we're building a richtext editor using CKEditor 5. We're calling `$dispatch('input', editor.getData())` inside CKEditors update callback function to dispatch the raw HTML string. This works very well until the user is trying to save an empty string. Livewire ignores the falsy value and falls back to the `value` property of the underlying `textarea` element, which still is the initial text before the user started editing.

Our workaround for now is falling back to a space: `$dispatch('input', editor.getData() || ' ')`. 
This is ok for the moment but obviously we would like to dispatch either `null` or an actually empty string.

**EDIT**: Just found another workaround. I'm also updating the `textarea`'s value property before dispatching like this:
```js
// get CKeditor raw data
const data = editor.getData()

// update the underlying textarea `value`
editor.sourceElement.value = data

// call alpine.js' dispatch
$dispatch('input', data)
```

**Comment by adriaanzon** [comment]
I wanted to create a proper fix for this, so I tried reproducing the bug described in #4247, but I couldn't. Using Livewire 2.7.2 (before the change from #4247), I am able to auto-fill inputs with Safari and the values get picked up by Livewire.

@codywoolsey Are you able to chime in here and explain how to reproduce the Safari auto-fill bug in Livewire <= 2.7.2?

**Reply by codywoolsey** [reply]
For me, this was happening in production and was very simple to reproduce. Essentially, we had a basic login form with a livewire component. This was a very simple form that was supposed to determine if the user was using SSO or not, and would hide/show the password field appropriately. So basically it would just be a component with a basic `render` function with the text inputs wired to the component via `wire:model`.

From there, you would follow the steps outlined in #4247:

1. Autofill your credentials, this may work properly the first time.
2. Clear out one of the values (username or password).
3. Autofill your credentials again. Whichever input that you didn't initiate the autofill from will be recognized as an undefined value. The input will be cleared, and the property of the component will erroneously be set to "null".

It's important to note that since the release of 2.8.0, Mac OS Monterey was released. I'm not sure if Safari was updated as part of this release (or has been updated in a minor update since), but it may not be possible anymore to replicate this issue if using the current version of Safari. Apple has a habit of changing things like this without any concern for breaking changes 😉

**Reply by adriaanzon** [reply]
Ah, for some reason I overlooked your steps to reproduce it in #4247. I can indeed reproduce it when using username/password autofill specifically. 

Now it's unfortunate that Safari uses the same value for e.detail as the value i'm actually trying to dispatch (`null`).

I wonder if there is some way Safari's CustomEvents can be distinguished from the ones Alpine sends using `$dispatch()`. Here's them side by side:

| Safari CustomEvent | Alpine Custom event with `null` payload |
| --- | --- |
| <img width="330" alt="Screenshot 2022-03-05 at 00 36 41" src="https://user-images.githubusercontent.com/4326420/156856269-dbd3a2f1-b921-4fca-ba74-95c5dfc7588a.png"> | <img width="330" alt="Screenshot 2022-03-05 at 00 36 21" src="https://user-images.githubusercontent.com/4326420/156856317-85881c13-d323-4965-8cf1-524c504c68b5.png"> |

Actually I remember listening to a podcast about this issue some time ago, it's [this one](https://podcasts.apple.com/nl/podcast/building-livewire/id1480884324?i=1000492969437).

**Comment by adriaanzon** [comment]
One workaround I found for allowing `$dispatch('input', null)`, is to use `$dispatch('input', 'NULL')` instead, and then in the Livewire component add an updating hook for the property. For example, for the `$buttonChoice` property from the [demo](https://github.com/adriaanzon/livewire-dispatch-bug-demo/blob/main/app/Http/Livewire/LivewireDispatchBugDemo.php#L9) in my original post:

```php
public function updatingButtonChoice(&$value): void
{
    if ($value === 'NULL') {
        $value = null;
    }
}
```

For falsy values other than null, I've created a fix in https://github.com/livewire/livewire/pull/4765.
